pkg/reconciler/common/prune.go

```diff
@@ -27,527 +27,775 @@ import (
+					continue
+				fi
+				if [[ "$resource_list" == "" ]]; then
+					echo "there is no '$parent_resource' available in '$namespace'"
+					continue
+				fi
+			fi
+			target_cmd="tkn $resource $resource_list delete $updated_flags --namespace=$namespace --force"
+			echo "\$ ${target_cmd}"
+			eval $target_cmd
+			if [ $? -ne 0 ]; then
+				error_status=1
+			fi
+		done
+		IFS=$old_ifs
+	}
+	
+	error_status=0
+	for c in $*; do
+		namespace=$(echo $c | cut -d ';' -f 1)
+		flags=$(echo $c | cut -d ';' -f 2)
+		resources=$(echo $c | cut -d ';' -f 3)
+		apply_to_individual_resource=$(echo $c | cut -d ';' -f 4)
+		prune $namespace $flags $resources $apply_to_individual_resource
+	done
+	exit $error_status
+	`
+)
+
+var (
+	// normalize resources
+	pruneResourceNameMap = map[string]string{
+		"pipelineruns": "pipelinerun",
+		"pipelinerun":  "pipelinerun",
+		"pr":           "pipelinerun",
+		"taskruns":     "taskrun",
+		"taskrun":      "taskrun",
+		"tr":           "taskrun",
+	}
 )
 
 type Pruner struct {
-	kc              kubernetes.Interface
+	tektonConfig    *v1alpha1.TektonConfig
+	kubeClientset   kubernetes.Interface
 	tknImage        string
 	targetNamespace string
-	ownerRef        v1.OwnerReference
+	ownerRef        metav1.OwnerReference
+	logger          *zap.SugaredLogger
+	ctx             context.Context
 }
 
-type pruneConfigPerNS struct {
-	config v1alpha1.Prune
+type pruneConfig struct {
+	Schedule                  string
+	Namespace                 string
+	Keep                      *uint
+	KeepSince                 *uint
+	Resources                 []string
+	ApplyToIndividualResource bool
+	TknImage                  string
 }
 
-// all the namespaces of default and the annotationbased
-type pruningNs struct {
-	commonScheduleNs map[string]*pruneConfigPerNS
-	uniqueScheduleNS map[string]*pruneConfigPerNS
-	configChanged    bool
+func Prune(ctx context.Context, k kubernetes.Interface, tektonConfig *v1alpha1.TektonConfig) error {
+	pruner, err := getPruner(ctx, k, tektonConfig)
+	if err != nil {
+		return err
+	}
+
+	return pruner.reconcile()
 }
 
-func Prune(ctx context.Context, k kubernetes.Interface, tC *v1alpha1.TektonConfig) error {
+func getPruner(ctx context.Context, k kubernetes.Interface, tektonConfig *v1alpha1.TektonConfig) (*Pruner, error) {
 	pruner := &Pruner{
-		kc:              k,
-		targetNamespace: tC.Spec.TargetNamespace,
-		ownerRef:        *v1.NewControllerRef(tC, tC.GetGroupVersionKind()),
+		tektonConfig:    tektonConfig,
+		kubeClientset:   k,
+		targetNamespace: tektonConfig.Spec.TargetNamespace,
+		ownerRef:        *metav1.NewControllerRef(tektonConfig, tektonConfig.GetGroupVersionKind()),
+		ctx:             ctx,
+		logger:          logging.FromContext(ctx),
+	}
+	return pruner, nil
+}
+
+func (pr *Pruner) reconcile() error {
+	// if an image specified in the tektonConfig spec, use that
+	if pr.tektonConfig.Spec.Pruner.Image != "" {
+		pr.tknImage = pr.tektonConfig.Spec.Pruner.Image
+	} else {
```

Comment:
This defaulting we can do at config CR level

Reply:
@piyush-garg could you please elaborate more on this?
Would you want to add the default image in to `tektonConfig.Spec.Pruner.Image` ?

Reply:
Yes, i mean the same, if it is something specified in CR then fine, otherwise take the default from there.

here it should just read the CR value

Reply:
@piyush-garg I reverted the image option via CR